### DISCUSSION TITLE
Convergence error in Cohesive zone model

### DISCUSSION THREAD

**Discussion by sidharthsarmah** [idaholab/moose#27189] [Q&A Modules: Solid mechanics]
Hi @arovinelli , I am trying to implement a simple cohesive zone model, but I encountered convergence issue on unloading. As suggested in paper (https://www.sciencedirect.com/science/article/pii/S0013794416303198), I can implement a viscous parameter to converge it, but I am still encountering convergence issue, could you suggest on how to overcome it.
Is it related to the value of the viscous parameter or something related to refinement of mesh, please suggest.

**Comment by sidharthsarmah** [comment]


**Reply by GiudGiud** [reply]
hello

no idea on this particular topic. I would go through the steps here
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

**Reply by sidharthsarmah** [reply]
Thanks @GiudGiud

**Reply by sidharthsarmah** [reply]
Hi @GiudGiud , I am getting this error, could you suggest. its working in my system but not in cluster:
```
*** ERROR ***
Unable to open file "/lustre04/scratch/sidharth/CZM_4340L/100_Notch.inp-mesh.cpr". Check to make sure that it exists and that you have read permission.
```

**Reply by GiudGiud** [reply]
what does `ls /lustre04/scratch/sidharth/CZM_4340L/` return?
MOOSE is expecting to see a mesh file there, looks like it s not there

**Reply by jiangwen84** [reply]
Is your new model different than the existing models in MOOSE?

In general, convergence issue is complicated. If you are using existing models, I would suggest you look at your problem setup, solver choice, loading conditions, etc. If you are developing your own model, then I would suggest check your Jacobian, residual.

**Reply by arovinelli** [reply]
Hi @sidharthsarmah, if you have issues on unloading (i.e. reducing the jump) you are probably not keeping track of a damage parameter properly or you haven't implemented the unload branch of your CZM properly. (Unloading has a different path than loading when damage is present)

**Reply by sidharthsarmah** [reply]
Hi @arovinelli , maybe I am using the wrong term 'unloading', I meant to say this part when the traction starts to drop to zero as separation occurs. There is issue related to convergence here. <img width="424" alt="image" src="https://github.com/idaholab/moose/assets/110927986/650d8e50-bebf-4361-ac93-3bdf1e78547a">

When I introduce a viscous term (in red), it converges but the values are high, could you please check what is wrong:

<img width="325" alt="image" src="https://github.com/idaholab/moose/assets/110927986/a0cd3c93-21c8-4d94-a967-3ab246930f9f">

```
#include "ScheiderViscous.h"

registerMooseObject("SolidMechanicsApp", ScheiderViscous);

InputParameters
ScheiderViscous::validParams()
{
  InputParameters params = CZMComputeLocalTractionTotalBase::validParams();
    params.addClassDescription("Scheider cohesive law");
    params.addRequiredParam<Real>("normal_gap_at_maximum_normal_traction","d0_normal");
    params.addRequiredParam<Real>("tangential_gap_at_maximum_shear_traction","d0_shear");
    params.addRequiredParam<Real>("maximum_normal_traction","T0_normal");
    params.addRequiredParam<Real>("maximum_shear_traction", "T0_shear");
    params.addParam<MaterialPropertyName>("c_tot", "total concentration");
    params.addParam<Real>("viscosity", "Viscosity");
  return params;
}

ScheiderViscous::ScheiderViscous(const InputParameters & parameters)
  : CZMComputeLocalTractionTotalBase(parameters),
    _delta_u0({getParam<Real>("normal_gap_at_maximum_normal_traction"),
               getParam<Real>("tangential_gap_at_maximum_shear_traction"),
               getParam<Real>("tangential_gap_at_maximum_shear_traction")}),
    _max_allowable_traction({getParam<Real>("maximum_normal_traction"),
                             getParam<Real>("maximum_shear_traction"),
                             getParam<Real>("maximum_shear_traction")}),
    _interface_displacement_jump_old(
    getMaterialPropertyOld<RealVectorValue>(_base_name + "interface_displacement_jump")),
    _c_tot(getADMaterialProperty<Real>("c_tot")),
    _viscosity(getParam<Real>("viscosity"))
{
}
```
```
traction_local(i) = _viscosity*(_max_allowable_traction(i)/_delta_u0(i))*((_interface_displacement_jump[_qp](i)- _interface_displacement_jump_old[_qp](i))/_dt) + ...
```
```
traction_jump_derivatives_local(i, j) = _viscosity*(_max_allowable_traction(i)/_delta_u0(i))*(1/_dt) +  ...
```

**Reply by arovinelli** [reply]
@sidharthsarmah your traction is simply a function of the current jump and does not depends of history. There isn't an irreversible damage variable, thus I can see how your model does have trouble converging but gets better when you add viscosity.
You should add an irreversible damage parameter and properly account for reloading scenarios. This will fix most if not all of your problems

**Reply by sidharthsarmah** [reply]
Hi @arovinelli , I did use the previous jump in the equation:

Is the damage parameter one way to get rid of convergence, as it was mentioned viscosity is another way to get rid of it, please suggest. Can you refer on where do I need to add the damage variable?
<img width="210" alt="image" src="https://github.com/idaholab/moose/assets/110927986/d828242f-a706-4e34-a758-9e5dc7573286">

```
_interface_displacement_jump_old(
    getMaterialPropertyOld<RealVectorValue>(_base_name + "interface_displacement_jump")),
```
```
traction_local(i) = _viscosity*(_max_allowable_traction(i)/_delta_u0(i))*((_interface_displacement_jump[_qp](i)- _interface_displacement_jump_old[_qp](i))/_dt) + ...
```

**Reply by arovinelli** [reply]
but you still recover damage if your displacement jump decreases, i.e. your stiffness does not degrades monotonically. if you unload (change the BC to unload) you should not go back on the same path as when you loaded.

**Reply by sidharthsarmah** [reply]
Ok @arovinelli , I have use three condition for the 3 regions:
<img width="438" alt="image" src="https://github.com/idaholab/moose/assets/110927986/257282ab-f151-47e7-9fa3-0f0e71a790e4">
```
if (d < d1)
region 1
if (d1 < d < d2)
region 2
if (d2 < d < d3)
region 3
```
But inserting unloading pattern do you mean, I include a condition for this?
<img width="438" alt="image" src="https://github.com/idaholab/moose/assets/110927986/283aaafe-34fd-48d7-be11-874586c3be51">

```
else if (d<d_old)
unloading line
```
What does damage variable do? Please suggest, I have not worked on this before.

**Comment by arovinelli** [comment]
yes, but you need to make sure you have zero traction at 0 jump, also you need to specify what is going to happen if the traction / jump is negative (usually negative jump you want the undamaged stiffness times a penalty factor e.g. 1e3) which for now you can set to one

**Reply by sidharthsarmah** [reply]
Sorry @arovinelli , I am a bit confused about this terms you are mentioning, could you help me what changes I need to make in the code?

These are the two conditions which I have added to the TSL, please suggest.

<img width="578" alt="image" src="https://github.com/idaholab/moose/assets/110927986/6220f885-a8af-4562-9005-da76a68461ce">

```
if (_interface_displacement_jump[_qp](i) < 0.0)
                    traction_local(i) = _max_allowable_traction(i) * 1;
                
 if (_interface_displacement_jump[_qp](i) == 0.0)
                    traction_local(i) = _max_allowable_traction(i) * 0;

if (_interface_displacement_jump[_qp](i) < 0.01*_delta_u0(i))
                    traction_local(i) = _max_allowable_traction(i) * (2*(_interface_displacement_jump[_qp](i)/(0.01*_delta_u0(i))) - std::pow(_interface_displacement_jump[_qp](i)/(0.01*_delta_u0(i)), 2));
                
                else if (_interface_displacement_jump[_qp](i) >= 0.01*_delta_u0(i) && _interface_displacement_jump[_qp](i) <= 0.5*_delta_u0(i))
                    traction_local(i) = _max_allowable_traction(i);
                
                else if (_interface_displacement_jump[_qp](i) > 0.5*_delta_u0(i) && _interface_displacement_jump[_qp](i) <= 1*_delta_u0(i))
                    traction_local(i) = _viscosity*(_max_allowable_traction(i)/_delta_u0(i))*((_interface_displacement_jump[_qp](i)- _interface_displacement_jump_old[_qp](i))/_dt) + _max_allowable_traction(i) * (2*std::pow((_interface_displacement_jump[_qp](i)-0.5*_delta_u0(i))/(_delta_u0(i)-0.5*_delta_u0(i)), 3)  - 3*std::pow((_interface_displacement_jump[_qp](i)-0.5*_delta_u0(i))/(_delta_u0(i)-0.5*_delta_u0(i)), 2) + 1);
                
                else if (_interface_displacement_jump[_qp](i) > 1*_delta_u0(i))
                    traction_local(i) = 0.0000001*_interface_displacement_jump[_qp](i);
```

**Reply by arovinelli** [reply]
if delta <=0 T=T0*delta*penalty_factor
this is for the negative jump


the unloading conditions is a bit different a comes into play only for positive jumps
loading condition -> delta>0 & delta >= delta_old
unloading condition ->  delta>0 delta < delta_old

you will need to keep track of the maximum positive jump ever obtained, and its associated traction
delta_max_ever = max(delta_max_ever_old, current_delta)
T_delta_max_ever = water value comes from you undameged TSL for delta_max_ever

now assume damage kicks in only after delta_max_ever > delta1
if loading or if delta_max_ever_old <= delta 1
   T = whatever comes from your TSL for delta <1
else (unloading with delta_max_ever_old > delta1)
   T= delta/delta_max_ever*T_delta_max_ever
